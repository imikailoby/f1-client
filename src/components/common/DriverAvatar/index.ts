export { default as DriverAvatar } from './driverAvatar.component';
