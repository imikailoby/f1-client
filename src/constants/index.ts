export const API_URL: string = process.env.REACT_APP_API_URL || 'http://ergast.com/api/f1';
export const CURRENT_ENV: 'development' | 'production' | 'test' = process.env.NODE_ENV as
  | 'development'
  | 'production'
  | 'test';

// Value hardcoded because we cannot time traveling and start first F1 season before 1950;
export const FIRST_F1_SEASON_YEAR = 1950;
export const FROM_F1_SEASON_YEAR = Number(process.env.REACT_APP_FROM_F1_SEASON_YEAR || '2005');

export const WIKI_API_URL = 'https://en.wikipedia.org/w/api.php';
