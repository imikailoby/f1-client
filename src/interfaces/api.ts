import { MRData } from '../types/mrData';
import { RaceTable } from './raceTable';
import { StandingsListItem } from './standingsLists';

export interface IGetDriverStandingsResponse {
  MRData: MRData<{ StandingsTable: { StandingsLists: StandingsListItem[] } }>
}

export interface IGetResultsByYearResponse {
  MRData: MRData<{ RaceTable: RaceTable }>
}

export interface IGetDriverStandings {
  standingsLists: StandingsListItem[],
  pagination: {
    limit: MRData['limit'];
    offset: MRData['offset'];
    total: MRData['total'];
  }
}
