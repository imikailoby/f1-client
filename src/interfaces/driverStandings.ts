import { Constructor } from './constructor';
import { Driver } from './driver';

export interface DriverStandingsItem {
  Constructors: Constructor[];
  Driver: Driver,
  points: string;
  positionText: string;
  wins: string;
}
