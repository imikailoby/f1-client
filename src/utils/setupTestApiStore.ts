/* eslint-disable @typescript-eslint/no-explicit-any */
import { AnyAction, configureStore, EnhancedStore, Middleware, Reducer } from '@reduxjs/toolkit';

export function setupTestApiStore<
  A extends {
    reducer: Reducer<any, any>;
    reducerPath: string;
    middleware: Middleware;
    util: { resetApiState(): any };
  },
>(api: A) {
  const getStore = () =>
    configureStore({
      reducer: {
        [api.reducerPath]: api.reducer,
      },
      middleware: (gdm) => gdm({ serializableCheck: false, immutableCheck: false }).concat(api.middleware),
    });

  type StoreType = EnhancedStore<
    {
      api: ReturnType<A['reducer']>;
    },
    AnyAction,
    ReturnType<typeof getStore> extends EnhancedStore<any, any, infer M> ? M : never
  >;

  const initialStore = getStore() as StoreType;
  const refObj = {
    api,
    store: initialStore,
  };

  return refObj;
}
