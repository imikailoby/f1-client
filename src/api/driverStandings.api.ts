import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { IGetDriverStandings, IGetDriverStandingsResponse } from '../interfaces/api';
import { API_URL } from '../constants';
import { transformGetDriverStandingsApiResponse } from '../helpers/api.helpers';

export const driverStandingsApi = createApi({
  reducerPath: 'driverStandings',
  baseQuery: fetchBaseQuery({ baseUrl: API_URL }),
  tagTypes: ['DriverStandings'],
  endpoints: (builder) => ({
    getDriverStandings: builder.query<IGetDriverStandings, { limit: number, offset: number }>({
      query: (params) => ({ url: 'driverstandings/1.json', params }),
      providesTags: ['DriverStandings'],
      transformResponse: (response: IGetDriverStandingsResponse) => transformGetDriverStandingsApiResponse(response),
    }),
  }),
});

export const { useGetDriverStandingsQuery } = driverStandingsApi;
