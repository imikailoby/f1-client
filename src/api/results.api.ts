import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { IGetResultsBySeasonResponse } from '../interfaces/api';
import { API_URL } from '../constants';
import { RaceTable } from '../interfaces/raceTable';
import { transformGetResultsBySeasonResponse } from '../helpers/api.helpers';
import { StandingsListItem } from '../interfaces/standingsLists';

export const resultsApi = createApi({
  reducerPath: 'results',
  baseQuery: fetchBaseQuery({ baseUrl: API_URL }),
  tagTypes: ['Results'],
  endpoints: (builder) => ({
    getResultsBySeason: builder.query<RaceTable, { season: StandingsListItem['season'] }>({
      query: ({ season }) => `${season}/results/1.json`,
      providesTags: ['Results'],
      transformResponse: (response: IGetResultsBySeasonResponse) => transformGetResultsBySeasonResponse(response),
    }),
  }),
});

export const { useGetResultsBySeasonQuery } = resultsApi;
