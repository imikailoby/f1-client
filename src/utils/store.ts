import { configureStore } from '@reduxjs/toolkit';
import { driverStandingsApi } from '../api/driverStandings.api';
import { resultsApi } from '../api/results.api';
import { CURRENT_ENV } from '../constants';

const apiMiddlewares = [driverStandingsApi.middleware, resultsApi.middleware];

export const store = configureStore({
  reducer: {
    [driverStandingsApi.reducerPath]: driverStandingsApi.reducer,
    [resultsApi.reducerPath]: resultsApi.reducer,
  },
  devTools: CURRENT_ENV === 'development',
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(apiMiddlewares),
});

export type RootState = ReturnType<typeof store.getState>;
