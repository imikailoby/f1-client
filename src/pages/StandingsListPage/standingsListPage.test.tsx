/**
 * @jest-environment jsdom
 */
import { rest } from 'msw';
import { render, screen, act } from '@testing-library/react';
import { Provider } from 'react-redux';
import { server } from '../../../mocks/server';
import { driverStandingsApi } from '../../api/driverStandings.api';
import getDriverStandingsReponse from '../../../mocks/responses/getDriverStandingsReponse.json';
import { API_URL, WIKI_API_URL } from '../../constants';
import { StandingsListPage } from '.';
import { clearSessionStorage } from '../../utils/common';
import { setupTestApiStore } from '../../utils/setupTestApiStore';
import React from 'react';

const storeRef = setupTestApiStore(driverStandingsApi);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => {
  server.close();
  clearSessionStorage();
});

const MOCKED_PARAMS = { limit: 5, offset: 0 };

describe('<StandingsListPage />', () => {
  test('renders with loading state without crashing', () => {
    server.use(
      rest.get(`${API_URL}/driverstandings/1.json`, (_, res, ctx) => res(ctx.status(500))),
      rest.get(`${WIKI_API_URL}*`, (_, res, ctx) => res(ctx.status(500))),
    );

    const { container } = render(
      <Provider store={storeRef.store}>
        <StandingsListPage />
      </Provider>,
    );

    expect(screen.queryByAltText('loader')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('renders with results without crashing', async () => {
    server.use(
      rest.get(`${API_URL}/driverstandings/1.json`, (_, res, ctx) =>
        res(ctx.status(200), ctx.json(getDriverStandingsReponse)),
      ),
      rest.get(`${WIKI_API_URL}*`, (_, res, ctx) => res(ctx.status(500))),
    );

    const { container } = render(
      <Provider store={storeRef.store}>
        <StandingsListPage />
      </Provider>,
    );

    await act(async () => {
      storeRef.store.dispatch(driverStandingsApi.endpoints.getDriverStandings.initiate(MOCKED_PARAMS));
    });

    expect(screen.queryByAltText('loader')).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('update standings list state when new data received', async () => {
    server.use(
      rest.get(`${API_URL}/driverstandings/1.json`, (_, res, ctx) =>
        res(ctx.status(200), ctx.json(getDriverStandingsReponse)),
      ),
      rest.get(`${WIKI_API_URL}*`, (_, res, ctx) => res(ctx.status(500))),
    );

    const setStateMock = jest.fn();
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const useStateMock: any = (useState: any) => [useState, setStateMock];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);

    render(
      <Provider store={storeRef.store}>
        <StandingsListPage />
      </Provider>,
    );

    expect(setStateMock).toHaveBeenCalledTimes(0);

    await act(async () => {
      storeRef.store.dispatch(driverStandingsApi.endpoints.getDriverStandings.initiate(MOCKED_PARAMS));
    });

    expect(setStateMock).toHaveBeenCalledTimes(1);
  });
});
