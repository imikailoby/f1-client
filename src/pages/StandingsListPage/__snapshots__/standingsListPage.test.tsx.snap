// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StandingsListPage /> renders with loading state without crashing 1`] = `
<div>
  <div
    class="loader"
  >
    <div
      class="loader"
    >
      <img
        alt="loader"
        src="svg>
"
      />
    </div>
  </div>
</div>
`;

exports[`<StandingsListPage /> renders with results without crashing 1`] = `
<div>
  <div
    class="list"
  >
    <div
      style="overflow: visible; height: 0px; width: 0px;"
    />
    <div
      class="resize-triggers"
    >
      <div
        class="expand-trigger"
      >
        <div
          style="width: 1px; height: 1px;"
        />
      </div>
      <div
        class="contract-trigger"
      />
    </div>
  </div>
</div>
`;
